Sally7/ValueConversion/BufferHelper.cs

```diff
@@ -0,0 +1,109 @@
+﻿using Sally7.Internal;
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Sally7.ValueConversion;
+
+internal class BufferHelper
+{
+    public static int CopyBytes(ReadOnlySpan<byte> input, Span<byte> output, int numberOfBytes)
```

Comment:
For the product code why not just `input.CopyTo(output)`? This is a general purpose optimized version within .NET.
Is the specialized version of copy here as we know that only small sizes will be copied, etc.?


BTW: `numberOfBytes` should be given implicit via the spans (i.e. they should be sliced before calling this method).

Reply:
> For the product code why not just `input.CopyTo(output)`? This is a general purpose optimized version within .NET. Is the specialized version of copy here as we know that only small sizes will be copied, etc.?

No I just messed up I guess. I started with one copy method and I had the idea `input.CopyTo(output)` is slow, because it doesn't do anything optimized. I once read about copying optimizations but I don't know what they applied to and I couldn't find that information anymore (and didn't spend much time searching, or checking the implementation in dotnet/runtime). However the perceived slowness might very well be related to the data I passed in. I should've benchmarked this as well, and if there are gains in small optimizations I can still do a different implementation than the one in the BCL, but preferably I just use what's already there. From net8 onwards there's also optimized versions of `BinaryPrimitives.ReverseEndianness` for `span` (the dotnet PR referenced above that you also reviewed), I want to use those as well when available.

> BTW: `numberOfBytes` should be given implicit via the spans (i.e. they should be sliced before calling this method).

Another mishap during refactoring, so yes, you're absolutely right (and they should also be properly sliced in current Sally7 code, I guess only benchmarks are an exemption right now).

Reply:
`Span<T>.CopyTo` in .NET needs to be general purpose, i.e. can handle all given length as optimized as possible.
This -- of course -- induces some length check and a bigger non inlineable method.

If user code knows that input is always below a specific size or has constant size, then it's quite easy to write a CopyTo that is faster -- but also specialized for that use case.

I'd go with the builtin method, as it's well tested and safe. Then only switch to a custom version if benchmarks show it's worth it.
(Just writing the last sentence for completeness, as I'm aware we have the same thinking here)

Reply:
I added a benchmark comparing my implementation against `Span.CopyTo`, the latter was faster for arrays with power of two lengths, so I just favored `Span.CopyTo`. There's room for improvement in mine, but I don't think that would actually turn the tables.

Reply:
> for arrays with power of two lengths

Yep, it's due to copying via `short` (2), `int` (4), `long` (8), `Vector128<byte>` (16), `Vector256<byte>` (32).

But thanks for measuring.

Reply:
> > for arrays with power of two lengths
> 
> Yep, it's due to copying via `short` (2), `int` (4), `long` (8), `Vector128<byte>` (16), `Vector256<byte>` (32).

Which is exactly what I was doing, but only up to long 😉